File: samples-java/src/main/java/io/github/ilaborie/loops/samples/recursion/Accumulate.java
package io.github.ilaborie.loops.samples.recursion;

import io.github.ilaborie.loops.samples.Element;

import java.util.List;

import static io.github.ilaborie.loops.samples.Elements.getSomeElements;

public class Accumulate {

    public static void main(String[] args) {
        List<Element> elements = getSomeElements();
        int result = accumulate(elements);
        System.out.println(result);
    }

    public static int accumulate(List<Element> elements) {
        if (elements.isEmpty()) {
            return 0;
        }

        Element head = elements.get(0);
        List<Element> tail = elements.subList(1, elements.size());

        int count = head.getName().length();
        int countTail = accumulate(tail);
        return count + countTail;
    }

    public static int accumulateAlt(List<Element> elements) {
        if (elements.isEmpty()) {
            return 0;
        }

        Element head = elements.get(0);
        List<Element> tail = elements.subList(1, elements.size());
        return head.getName().length() + accumulateAlt(tail);
    }

}

File: samples-kotlin/src/main/kotlin/io/github/ilaborie/loops/samples/recursion/accumulate.kt
package io.github.ilaborie.loops.samples.recursion

import io.github.ilaborie.loops.samples.ElementKt
import io.github.ilaborie.loops.samples.stream.accumulate
import io.github.ilaborie.loops.samples.getSomeElementsKt


fun main() {
    val elements = getSomeElementsKt()
    val result = accumulate(elements)
    println(result)
}

fun accumulate(elements: List<ElementKt>): Int {
    if (elements.isEmpty()) {
        return 0
    }

    val head = elements.first()
    val tail = elements.drop(1)

    val count = head.name.length
    val countTail = accumulate(tail)
    return count + countTail
}

fun accumulateAlt(elements: List<ElementKt>): Int =
    if (elements.isEmpty()) 0
    else elements.first().name.length + accumulateAlt(elements.drop(1))
